[EvitanRelta] iP

EvitanRelta opened the pull request:
# DukePro

> "Your mind is for having ideas, not holding them.” – David Allen ([source])

[source]: https://dansilvestre.com/productivity-quotes

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here][download].
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is FREE!

[download]: https://nus-cs2103-ay2223s2.github.io/website/scedule/week4/project.html

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you're a `Java` programmer, you can use it to practice Java too. Here's a random `main` method that has nothing to do with my `Duke`:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

EvitanRelta committed:
Abstract out printing text to console

EvitanRelta committed:
Add echo & exit commands

EvitanRelta committed:
Add indentation to console output

EvitanRelta committed:
Add adding feat. & list command

EvitanRelta committed:
Replace task-list array with classes

EvitanRelta committed:
Abstract out the parsing of commands into a class

EvitanRelta committed:
Abstract commands to individual methods

EvitanRelta committed:
Add mark & unmark commands

EvitanRelta committed:
Rearrange method declarations

EvitanRelta committed:
Create todo, event & deadline task classes

EvitanRelta committed:
Add get status method to tasklist

EvitanRelta committed:
Add todo command

EvitanRelta committed:
Rename private variables in 'Command' class

EvitanRelta committed:
Overhaul 'Command' class

EvitanRelta committed:
Add 'deadline' & 'event' commands

EvitanRelta committed:
Handle no-input scenerios

EvitanRelta committed:
Update base text-ui-test to fit output

EvitanRelta committed:
Add header message to 'list' command

EvitanRelta committed:
Add tests for 'Level-3' feats.

EvitanRelta committed:
Add tests for 'Level-4' feats.

EvitanRelta committed:
Trim command & command-param bodies

EvitanRelta committed:
Replace default 'namedParameter' value from 'null' to empty str.

EvitanRelta committed:
Change handling of unknown commands to give error

EvitanRelta committed:
Handle errors for 'mark' & 'unmark' commands

EvitanRelta committed:
Move handling of empty task-list to 'TaskList' class

EvitanRelta committed:
Remove the now invalid 'Level-3' tests

EvitanRelta committed:
Add tests for unknown commands

EvitanRelta committed:
Add tests for invalid 'mark' & 'unmark' usage

EvitanRelta committed:
Add test for '/to' event param b4 '/from'

EvitanRelta committed:
Handle errors for 'todo', 'deadline' & 'event' commands

EvitanRelta committed:
Add 'delete' command

EvitanRelta committed:
Implement saving/loading tasklist feat.

EvitanRelta committed:
Replace 'String' by/from/to task dates to 'LocalDate'

EvitanRelta committed:
Change task dates' print format

EvitanRelta committed:
Handle invalid dates for deadlines & events

EvitanRelta committed:
Merge branch 'branch-Level-7'

EvitanRelta committed:
Merge branch 'branch-Level-8'

EvitanRelta committed:
Merge branch 'add-gradle-support'

EvitanRelta committed:
Prevent termination on tasklist load error

EvitanRelta committed:
Split 'Duke.main' into a constructor & 'run' method

EvitanRelta committed:
Convert static 'say' method to instance method

EvitanRelta committed:
Convert all static Duke command methods to instance methods

EvitanRelta committed:
Abstract out ui-related code into a class

EvitanRelta committed:
Abstract out save/load-related code into a class

EvitanRelta committed:
Rename 'Command' class to 'Parser'

EvitanRelta committed:
Rename 'DukeLoadException' to 'DukeSaveLoadException'

EvitanRelta committed:
Handle error writing to save file

EvitanRelta committed:
Move command-line scanning code to 'Ui' class

EvitanRelta committed:
Move all files into 'duke' package

EvitanRelta committed:
Group exception & task classes into their own packages

EvitanRelta committed:
Make 'DukeException' message-prefix static

EvitanRelta committed:
Abstract out invalid-command message as a constant

EvitanRelta committed:
Gitignore '.vscode' dir

EvitanRelta committed:
Add tests for Duke exceptions

EvitanRelta committed:
Implement 'equals' method for tasks

EvitanRelta committed:
Add tests for tasks

EvitanRelta committed:
Fix gradle 'mainClassName' config

EvitanRelta committed:
Add JavaDocs

EvitanRelta committed:
Remove 'temp' variable namings

EvitanRelta committed:
Add spacing after 'catch' keyword

EvitanRelta committed:
Remove indents for switch cases

EvitanRelta committed:
Add empty line between methods

EvitanRelta committed:
Fix JavaDocs coding standard violations

EvitanRelta committed:
Adhere to wrapped code line coding standard

EvitanRelta committed:
Add constructors to 'TaskList'

EvitanRelta committed:
Fix 'runtext.bat' compilation paths

EvitanRelta committed:
Move 'list' command's header from 'TaskList' to 'Duke'

EvitanRelta committed:
Add 'find' command

EvitanRelta committed:
Merge branch 'branch-A-JavaDoc'

EvitanRelta committed:
Merge branch 'branch-A-CodingStandard'

EvitanRelta committed:
Merge branch 'branch-Level-9'

EvitanRelta committed:
Add 'q' command as alias for 'quit'

EvitanRelta committed:
Setup checkstyle

EvitanRelta committed:
Add JavaFX dependencies

WeideTeo commented on src/test/java/duke/exceptions/DukeInvalidCommandExceptionTest.java:
```diff
@@ -0,0 +1,23 @@
+
+public class DukeInvalidCommandExceptionTest {
+    @Test
+    public void getMessage() {
```
Will be good if can include JavaDocs for this class 👍

Lava-Iris commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,291 @@
+                + "| | | | | | | |/ / _ \\\n"
+                + "| |_| | |_| |   <  __/\n"
+                + "|____/ \\__,_|_|\\_\\___|\n";
+        this.ui.show("Hello from\n" + logo);
```
Maybe this step can be extracted out as a method like greet?

Lava-Iris commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,291 @@
+                case "exit":
+                case "bye":
+                    this.ui.show("Bye. Hope to see you again soon!");
+                    break whileLoop;
```
I like how you used the properties of switch to add multiple ways to say a command. It didn't occur to me we could do this.

Lava-Iris commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,291 @@
+     * 
+     * @throws DukeInvalidArgumentException If no description given.
+     */
+    private void addTodo() throws DukeInvalidArgumentException {
```
Maybe the implementation of the commands can be abstracted out as their own classes under a command super class?

Lava-Iris commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,130 @@
+     * @param values The array of values to encode.
+     * @return The encoded values.
+     */
+    protected static String encodeValues(String[] values) {
```
I really like this idea. I didn't think about the implications of using | in encoding.

Lava-Iris reviewed (commented):
Hi!

I learned some things I hadn't considered before from your code, so thank you for that. I couldn't find any major problems. Your testing was extensive and OOP was well implemented. 

All the best for future increments!

WeideTeo commented on src/test/java/duke/tasks/TaskTodoTest.java:
```diff
@@ -0,0 +1,36 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TaskTodoTest {
+    private static final String TEST_UNESCAPED_DESC = "VERTICAL | BAR";
```
Really like the idea of DEFINE the string rather than writing it out every time. Neat and clean

WeideTeo reviewed (commented):
Looks Good To me. Really did an excellent job of using OOP . Thumbs up !!

EvitanRelta committed:
Fix all checkstyle violations

EvitanRelta committed:
Auto-format all java files

EvitanRelta committed:
Relax trailing whitespace checkstyle for JavaDocs

EvitanRelta committed:
Implement JavaFX tut. part 1

EvitanRelta committed:
Implement JavaFX tut. part 2

EvitanRelta committed:
Implement JavaFX tut. part 3 iteration 1-2

EvitanRelta committed:
Fix Gradle main classname config

EvitanRelta committed:
Implement JavaFX tut. part 3 iteration 3

EvitanRelta committed:
Implement JavaFX tut. part 4

EvitanRelta committed:
Integrate Duke with the new JavaFX GUI

EvitanRelta committed:
Change dialog font to monospace font

EvitanRelta committed:
Prevent dialog text truncation

Based on the suggestion:
https://github.com/nus-cs2103-AY2223S2/forum/issues/123

EvitanRelta committed:
Implement Duke quiting command

EvitanRelta committed:
fixup! Implement JavaFX tut. part 4

EvitanRelta committed:
Merge branch 'branch-A-CheckStyle'

EvitanRelta committed:
Merge branch 'branch-Level-10'

EvitanRelta committed:
Fix command body parsing

The user input "deadline /by" fails to extract the empty command body,
and the 'by' parameter is improperly parsed due to the empty body.

It needs to be fixed so that Duke can detect the empty body, and give
the proper error message.

Let's update the Regex that's used to split the body from the
parameters, excluding the space before the param slash '/'.

By splitting on the slash, it would properly obtain the empty body,
should the body be empty.

EvitanRelta committed:
Add Java assertions

EvitanRelta committed:
Fix header comment of JavaDoc

As per: https://github.com/EvitanRelta/ip/issues/1

EvitanRelta committed:
Remove unnecessary use of 'this'

EvitanRelta committed:
Replace 'null' exit response with constant

To follow the spirit of the "Avoid Magic Numbers" code quality.

EvitanRelta committed:
Rename params & fields to be more descriptive

EvitanRelta committed:
Merge pull request #2 from EvitanRelta/branch-A-Assertions

Add Java assertions

EvitanRelta committed:
Merge branch 'master' into branch-A-CodeQuality

EvitanRelta committed:
Merge pull request #3 from EvitanRelta/branch-A-CodeQuality

Improve code quality

EvitanRelta committed:
Add 'sort' command

EvitanRelta committed:
Merge branch 'branch-C-Sort'

EvitanRelta committed:
Fix parsing of command params

EvitanRelta committed:
Add a representative screenshot

EvitanRelta committed:
Update user guide
